kotlin-native/Interop/Runtime/src/native/kotlin/kotlinx/cinterop/NativeMem.kt

```diff
@@ -22,91 +21,119 @@ internal external fun getPointerSize(): Int
+    external fun putByte(mem: NativePointed, value: Byte)
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getShort(mem: NativePointed): Short
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putShort(mem: NativePointed, value: Short)
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getInt(mem: NativePointed): Int
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putInt(mem: NativePointed, value: Int)
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getLong(mem: NativePointed): Long
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putLong(mem: NativePointed, value: Long)
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getFloat(mem: NativePointed): Float
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putFloat(mem: NativePointed, value: Float)
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getDouble(mem: NativePointed): Double
+
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putDouble(mem: NativePointed, value: Double)
 
-    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE) external fun getShort(mem: NativePointed): Short
-    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE) external fun putShort(mem: NativePointed, value: Short)
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getNativePtr(mem: NativePointed): NativePtr
 
-    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE) external fun getInt(mem: NativePointed): Int
-    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE) external fun putInt(mem: NativePointed, value: Int)
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putNativePtr(mem: NativePointed, value: NativePtr)
 
-    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE) external fun getLong(mem: NativePointed): Long
-    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE) external fun putLong(mem: NativePointed, value: Long)
+    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE)
+    external fun getVector(mem: NativePointed): Vector128
 
-    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE) external fun getFloat(mem: NativePointed): Float
-    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE) external fun putFloat(mem: NativePointed, value: Float)
+    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE)
+    external fun putVector(mem: NativePointed, value: Vector128)
 
-    @TypedIntrinsic(IntrinsicType.INTEROP_READ_PRIMITIVE) external fun getDouble(mem: NativePointed): Double
-    @TypedIntrinsic(IntrinsicType.INTEROP_WRITE_PRIMITIVE) external fun putDouble(mem: NativePointed, value: Double)
+    @TypedIntrinsic(IntrinsicType.INTEROP_SET_MEMORY)
+    external fun memset(mem: NativePointed, value: Byte, size: Int)
```

Comment:
From what i know about LLVM, they should be equally portable.
However, there's a reason why i split them up:

All major native platforms apart from some mobile and embedded applications are 64-bit now.
Using 64-bit everywhere for things like memcpy could waste CPU cycles for applications where 64-bits will never be used (since this affects the internal index size of the intrinsic implementation afaik), like when copying between native memory <-> kotlin arrays, only if it's probably very little.
It is still important to offer the 64-bit API to allow utilizing the full capability of the host CPU when working with unmanaged memory.

If you want, we can drop the small 32-bit optimization entirely and just cast to 64-bit size types when dealing with things like arrays etc.

Reply:
I could be wrong to be honest, will look into this a little deeper.

Reply:
> All major native platforms apart from some mobile and embedded applications are 64-bit now.

Unfortunately, we still have [32-bit targets](https://kotlinlang.org/docs/native-target-support.html), and providing 64-bit API for them might be wrong.

Reply:
Then we should offer both, since Kotlin does not have a size type i think. 32-bit only is very mediocre for more serious applications imo.